File: Java/프로그래머스/2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.java
class Solution {
    int[][] matrix;
    public int[] solution(int rows, int columns, int[][] queries) {
        int[] answer = new int[queries.length];
        matrix = new int[rows][columns];
        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[0].length; j++) {
                matrix[i][j] = i * matrix[0].length + j + 1;
            }
        }

        for (int i = 0; i < queries.length; i++) {
            int[] range = queries[i];
            answer[i] = rotate(matrix, range);
        }

        return answer;
    }

    int rotate(int[][] matrix, int[] range) {
        int startRow = range[0] - 1;
        int endRow = range[2] - 1;
        int startCol = range[1] - 1;
        int endCol = range[3] - 1;

        int tmp = matrix[startRow][startCol];
        int min = Integer.MAX_VALUE;

        for (int i = startCol + 1; i <= endCol; i++) {
            int current = matrix[startRow][i];
            min = Math.min(min, Math.min(tmp, current));
            matrix[startRow][i] = tmp;
            tmp = current;
        }

        for (int i = startRow + 1; i <= endRow; i++) {
            int current = matrix[i][endCol];
            min = Math.min(min, Math.min(tmp, current));
            matrix[i][endCol] = tmp;
            tmp = current;
        }

        for (int i = endCol - 1; i >= startCol; i--) {
            int current = matrix[endRow][i];
            min = Math.min(min, Math.min(tmp, current));
            matrix[endRow][i] = tmp;
            tmp = current;
        }

        for (int i = endRow - 1; i >= startRow; i--) {
            int current = matrix[i][startCol];
            min = Math.min(min, Math.min(tmp, current));
            matrix[i][startCol] = tmp;
            tmp = current;
        }

        return min;
    }
}

File: Python/프로그래머스/2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.py
def solution(rows, columns, queries):
    answer = [0] * len(queries)
    matrix = [[i * columns + j + 1 for j in range(columns)] for i in range(rows)]

    for i, query in enumerate(queries):
        answer[i] = rotate(matrix, query)

    return answer


def rotate(matrix, query):
    row_s, row_e, col_s, col_e = query[0] - 1, query[2] - 1, query[1] - 1, query[3] - 1
    tmp = matrix[row_s][col_s]
    min_num = float("INF")

    for i in range(col_s + 1, col_e + 1):
        cur = matrix[row_s][i]
        min_num = min(min_num, min(tmp, cur))
        matrix[row_s][i] = tmp
        tmp = cur

    for i in range(row_s + 1, row_e + 1):
        cur = matrix[i][col_e]
        min_num = min(min_num, min(tmp, cur))
        matrix[i][col_e] = tmp
        tmp = cur

    for i in range(col_e - 1, col_s - 1, -1):
        cur = matrix[row_e][i]
        min_num = min(min_num, min(tmp, cur))
        matrix[row_e][i] = tmp
        tmp = cur

    for i in range(row_e - 1, row_s - 1, -1):
        cur = matrix[i][col_s]
        min_num = min(min_num, min(tmp, cur))
        matrix[i][col_s] = tmp
        tmp = cur

    return min_num